### DISCUSSION TITLE
eXo openId configuration redirectUrl getting Issue

### DISCUSSION THREAD

**Discussion by Rajamuthukrishna** [exoplatform/exo-community#44] [Q&A]
Hi Team,

We are trying to access the eXo application in some other application, while try to hit the api getting CORS and 302 Issue,

Regards,
Raja

**Comment by rdenarie** [comment]
Hello @Rajamuthukrishna 

You will have to configure the nginx to add the header Allow-Access-control-origin on all requests : 
https://serverfault.com/questions/162429/how-do-i-add-access-control-allow-origin-in-nginx

Update the file ./conf/nginx.conf to add such header.

**Reply by Rajamuthukrishna** [reply]
Hello @rdenarie ,

As suggested, I have updated the headers in nginx, getting the below exception, please find the logs and help to resolve the issue.
[nginx.txt](https://github.com/exoplatform/exo-community/files/12892645/nginx.txt)

<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/40d4cd4a-6381-4d6a-9962-f854e4b0101d">

Regards,
Raja

**Reply by rdenarie** [reply]
Hello Raja,

I'm not nginx expert, so I will not be able to help you on this point. 

What I see is not related to CORS, so the configuration seems to work. 
The first problem is coming from the browser, which that your are trying to load a http resource in a https page, and the browser blocks it.

The second one seems not related to eXo

The 2 last are related to ws, but as I understand you tried to access to eXo API, it should not be a problem.

**Reply by Rajamuthukrishna** [reply]
Hello @rdenarie 
We are trying to access the eXo url in some other application. getting the 302 exception, kindly help to resolve the issue.
<img width="865" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/ef11fee9-15b7-47e2-b260-225014790a65">

**Comment by rdenarie** [comment]
Hello,
HTTP 302 is not an exception. 
It is a redirect code, so the application redirects you to /portal/sso?initialURI=%2Fportal%2Fdw%2F

It means that your request requires an authentication, and it redirect you the SSO authentication workflow

**Reply by Rajamuthukrishna** [reply]
Hello @rdenarie x

Thanks for your prompt response.

Kindly share the documentation or videos for configuring eXo SSO configuration with keyclock using openId.

**Comment by rdenarie** [comment]
Hi,
You will find documentation related to eXo Platform configuration for OpenId here : https://docs.exoplatform.org/administration/oauth-integration.html#registering-your-oauth-application

**Reply by Rajamuthukrishna** [reply]
Hi @rdenarie 
As you suggest the docs for OpenId Configuration - I have configured Keyclock and exo properties but facing below issue, kindly let me know if any configuration is missed. please help me resolve the issue.
[exo.txt](https://github.com/exoplatform/exo-community/files/13186386/exo.txt)
<img width="906" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/93ea1b12-1a9c-47bb-95d6-d745eb592116">
<img width="581" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/39392e63-b975-45ab-856c-a8fbd740f7f0">
<img width="593" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/94182e08-5138-4627-825a-24a622461def">
<img width="578" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/b3931c1e-285c-448e-ab58-07da680db215">
<img width="451" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/959d0a7a-6bc8-4f81-b69d-b7be18e393e3">
<img width="733" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/3b566cb1-d0cc-4f9f-bac9-0ee14456b265">
<img width="832" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/25279c17-8fb8-4321-8abb-9089fdc61cb7">
<img width="941" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/387dd1db-21dc-4603-8088-652bbe40e95e">

**Reply by rdenarie** [reply]
In exo.properties, the properties wellKnowConfigurationUrl and redirectUrl are commented.
Could you uncomment it ?

As you can see in your last screenshot, the redirect_uri is not replaced in the request url (we can see ${exO.oauth.openid.redirectURL}

**Reply by Rajamuthukrishna** [reply]
Hello @rdenarie 
I have updated the exo.properties and restart the service. please find the below logs and let me know if any configuration is missed...
<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/93859f11-1c6a-453a-b56d-3b835ee2840c">
<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/8f184461-e9c6-4b0c-a62c-aff60dabebc2">
[exo.txt](https://github.com/exoplatform/exo-community/files/13204259/exo.txt)

**Reply by rdenarie** [reply]
Please check your redirectURL configuration. 
It should be 
exo.oauth.openid.redirectURL=http://localhost/portal/openidAuth
instead of 
exo.oauth.openid.redirectURL=http://localhost/portal/login/openidAuth

**Reply by Rajamuthukrishna** [reply]
@rdenarie  - Have used the below configuration in exo.prop and tried getting some null in RequestUrl. please find the logs and kindly help to resolve the issue.
# OAuth
exo.base.url=http://localhost/portal/login
## OpenId
exo.oauth.openid.enabled=true
exo.oauth.openid.clientId=exo-openid
exo.oauth.openid.clientSecret=etR0yjUJ7fK9sNuobgvkxRgqywPHi9RG
## OpenId Urls
exo.oauth.openid.wellKnownConfigurationUrl=https://openid.server.com/.well-known/openid-configuration
exo.oauth.openid.redirectURL=http://localhost/portal/openidAuth
exo.oauth.registraterOnFly=OPENID
<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/88498b1f-0204-4daf-8d0d-2a1a6aea551c">

<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/93802e58-6b7d-4b30-ad52-e33451f3fd95">
<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/33b3879c-9d76-49e0-99c3-e1040b49064a">

<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/34d7bdae-8e10-4fb5-aa4f-331dc5422683">

**Comment by rdenarie** [comment]
This property seems also not correct :
exo.oauth.openid.wellKnownConfigurationUrl=https://openid.server.com/.well-known/openid-configuration
You must adapt it to have the well-known url of your IDP server.
You can find information on how to find it in keycloak here : https://stackoverflow.com/questions/28658735/what-are-keycloaks-oauth2-openid-connect-endpoints

For Using google as IDP for example, the url is : https://accounts.google.com/.well-known/openid-configuration

**Comment by rdenarie** [comment]
Can you provide platform log, so that I can check exactly where the problem is coming from ?
Thanks

**Reply by Rajamuthukrishna** [reply]
@rdenarie  - modified the exo.prop but still having the same null issue in redirectUrl. Please find the logs..
While click the signin with single account.,  first request is 
http://localhost/portal/openidAuth?_oauthInteraction=start&_initialURI=%2Fportal&_rememberme=true
then redirect to 
http://localhost/portal/null?response_type=code&client_id=exo-openid&scope=openid%20profile%20email&redirect_uri=http://localhost/portal/openidAuth&state=-2988670562276671434

[exo.txt](https://github.com/exoplatform/exo-community/files/13214601/exo.txt)
<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/63a64834-2108-445d-9995-dbed8864bae4">
<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/83f901f4-9bf4-4447-b05e-1c4e9dfc06c3">
<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/259180b1-4513-4488-bc10-f62eca6e00ff">

**Reply by Rajamuthukrishna** [reply]
@rdenarie  - Can you please find the logs and help to find the configuration generating the redirecturl causing issue..
**http://localhost/portal/null**?response_type=code&client_id=exo-openid&scope=openid%20profile%20email&redirect_uri=http://localhost/portal/openidAuth&state=1951980419119501194
Logs:
[eXo-logs.txt](https://github.com/exoplatform/exo-community/files/13226698/eXo-logs.txt)

**Reply by Rajamuthukrishna** [reply]
@rdenarie @plamarque @margondicco @ahamdi   - I have shared the all the configuration and properties files.  We are waiting for your response kindly find the details and help to resolve the configuration redirecting url issue.

Keyclock Configuration:
<img width="1018" alt="keyclock_01" src="https://github.com/exoplatform/exo-community/assets/112479162/f44c9928-aebd-400f-b51a-305ca3e3e25a">
<img width="960" alt="keyclock_02" src="https://github.com/exoplatform/exo-community/assets/112479162/048d0845-c1fc-4441-9ab4-b72bea9e5d8e">
<img width="960" alt="keyclock_03" src="https://github.com/exoplatform/exo-community/assets/112479162/f8c30d44-0b00-43a4-8c87-414785bc77a2">
<img width="960" alt="keyclock_04" src="https://github.com/exoplatform/exo-community/assets/112479162/38cf3171-01be-4141-b2c4-502d0e40a39e">
<img width="960" alt="keyclock_05" src="https://github.com/exoplatform/exo-community/assets/112479162/37ccdc71-c5cf-4bbc-9670-1870ebd4deff">
<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/909a585c-72f5-483d-8e6c-b244410b8977">
docker-compose:
[docker-compose.txt](https://github.com/exoplatform/exo-community/files/13239041/docker-compose.txt)
nginx.conf:
[nginx.txt](https://github.com/exoplatform/exo-community/files/13239049/nginx.txt)
exo.properties
[exo.txt](https://github.com/exoplatform/exo-community/files/13239058/exo.txt)
Logs:
[eXo-logs.txt](https://github.com/exoplatform/exo-community/files/13239302/eXo-logs.txt)

Redirecting URL having issue:
**http://localhost/portal/null**?response_type=code&client_id=exo-openid&scope=openid%20profile%20email&redirect_uri=http://localhost/portal/openidAuth&state=1951980419119501194

**Comment by rdenarie** [comment]
Hello, 
I'm sorry for the delay in the anwser, we are quite busy currently.

So, when looking in your logs, I can see the request /portal/openidAuth?_oauthInteraction=start&_initialURI=%2Fportal&_rememberme=true
This is call when you click on button "Connect with openId"

When click on this button, eXo will read the well-known configuration file, to search the entry 'authorization_endpoint'. Then the request redirect on this.

For example, I configure my local instance with google open id. 
The well-know url is https://accounts.google.com/.well-known/openid-configuration and you can see into  the 'authorization_endpoint' : 
  "authorization_endpoint": "https://accounts.google.com/o/oauth2/v2/auth",

When I click on the button on the login page, I see this : 
![image](https://github.com/exoplatform/exo-community/assets/807839/4d9ad296-bad1-4588-946e-e22bff917aaf)

The /portal/openidAuth (1) redirect to https://accounts.google.com/o/oauth2/v2/auth (2) with some parameters.

In your case, what is the value of authorization_endpoint in well-known configuration file ?

**Reply by Rajamuthukrishna** [reply]
Hello @rdenarie 
We have used the exo.properties and added the below configurations, but still having the issue, please suggest to resolve the issue.
[exo.txt](https://github.com/exoplatform/exo-community/files/13448580/exo.txt)

# OAuth
exo.base.url=http://localhost/portal/
## OpenId
exo.oauth.openid.enabled=true
exo.oauth.openid.clientId=exo-openid
exo.oauth.openid.clientSecret=Lv5apBzX3dRvtnMxKpFVDh6l6nGL2GLD
## OpenId Urls
exo.oauth.openid.wellKnownConfigurationUrl=http://localhost:8090/auth/realms/master/.well-known/openid-configuration
exo.oauth.openid.redirectURL=http://localhost:8090/auth/realms/master/o/oauth2/v2/auth
exo.oauth.registraterOnFly=OPENID
<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/a8e94d88-1f13-4af7-b554-41406072e46c">

Regards,
Raja

**Reply by rdenarie** [reply]
> In your case, what is the value of authorization_endpoint in well-known configuration file ?

You didnt provide this point

**Reply by Rajamuthukrishna** [reply]
@rdenarie  - Previously We are not used well-known configuration file, have used exo.properties file only and added the above configuration.
Application URLS:
eXo Application : http://localhost/portal/
Keyclock : http://localhost:8090/

As you suggested - have created the well-knownconfiguration.properties and added the below urls. please correct me..
## OpenId Urls
exo.oauth.openid.authorization_endpoint=http://localhost:8090/auth/realms/master/o/oauth2/v2/auth

**Reply by rdenarie** [reply]
well-know configuration is a file provided by the IDP. It contains all informations that exo need to use the IDP, it IS NOT a property on exo side.
The authorization_endpoint property is present inside the well-known configuration file.

When I use google as IDP for open ID, google provide the well-know configuration file, available here : https://accounts.google.com/.well-known/openid-configuration
As you can see in this file, the property is present.

Without this file, eXo CANNOT use open id protocol

Hint for keycloack : https://stackoverflow.com/questions/28658735/what-are-keycloaks-oauth2-openid-connect-endpoints

**Reply by Rajamuthukrishna** [reply]
@rdenarie  - Got it. We have used below wellKnownConfigurationUrl in exo.properties and authorization_endpoint also generated by the IDP using wellKnownConfigurationUrl but still that value is NOT picked BY exo.

exo.oauth.openid.wellKnownConfigurationUrl=http://localhost:8090/auth/realms/master/.well-known/openid-configuration

{
    "issuer": "http://localhost:8090/auth/realms/master",
    "authorization_endpoint": "http://localhost:8090/auth/realms/master/protocol/openid-connect/auth",
    "token_endpoint": "http://localhost:8090/auth/realms/master/protocol/openid-connect/token",
    "introspection_endpoint": "http://localhost:8090/auth/realms/master/protocol/openid-connect/token/introspect",
    "userinfo_endpoint": "http://localhost:8090/auth/realms/master/protocol/openid-connect/userinfo",
    "end_session_endpoint": "http://localhost:8090/auth/realms/master/protocol/openid-connect/logout",
    "frontchannel_logout_session_supported": true,
    "frontchannel_logout_supported": true,
    "jwks_uri": "http://localhost:8090/auth/realms/master/protocol/openid-connect/certs",
    "check_session_iframe": "http://localhost:8090/auth/realms/master/protocol/openid-connect/login-status-iframe.html",

<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/1c14813d-3690-468e-aa95-f95ecfc69971">

**Comment by rdenarie** [comment]
Do you use docker ? or standalone installation ?
We will need to add some trace logs to understand the problem.

Please, modify the file ${eXoHome}/conf/logback.xml by adding this line :
`  <logger name="io.meeds.oauth.openid" level="TRACE" />`
with other exiting loggers.
If you use docker, you will have to get the existing file inside the exo container, copy on the local machine, edit it, and modify the docker-compose file to mount it inside eXo
Then restart the platform, and try an access. 

Can you send the result log file ?

Thanks

**Reply by Rajamuthukrishna** [reply]
@rdenarie  - We have used docker only, In that container logback file is available,as you mentioned added the logger and restart the application, please find the logs file and configuration file - still having the issue.

[platform.log](https://github.com/exoplatform/exo-community/files/13467772/platform.log)

[logback.xml.txt](https://github.com/exoplatform/exo-community/files/13467776/logback.xml.txt)

[docker-compose.txt](https://github.com/exoplatform/exo-community/files/13467781/docker-compose.txt)

We have tried openId configuration using Google as IDP - Its working fine, keyclock IDP only having issue. Can you pls help us to resolve the issue.

Regards,
Raja

**Reply by rdenarie** [reply]
In your docker file, I cannot see how you mount the logback file.

Did you modify it inside the exo container and restart it ?
I'm not sure it is kept through different restart.
You must create this file on your local machine, and add line in exo -> volumes part

Please check in log file that you have TRACE logs before sending it.

Thanks

**Reply by Rajamuthukrishna** [reply]
@rdenarie  - docker-compose file - I have added the path( exo_logs:**/opt/exo/conf**)
<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/bf3d44db-7057-4cee-b9f4-1e66f8663a84">

logback.xml file have added the logger as you mentioned..
<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/026fed9b-8f65-421a-b518-2ba2f2afd3d4">

[exo.txt](https://github.com/exoplatform/exo-community/files/13473540/exo.txt)

both files are moved to container and restarted. Can you pls correct me - if anything is wrong..

**Reply by rdenarie** [reply]
> @rdenarie - docker-compose file - I have added the path( exo_logs:/opt/exo/conf)

The line with volume exo_logs must be kept. It is defining a docker volumes for logs. 
Please add a new line in volume part with this :

      - ./logback.xml:/opt/exo/conf/logback.xml:rw

in addition, the logback file must be in the same folder than the docker compose file.


Remember to send me the log file after a start, and a login attempt

**Reply by Rajamuthukrishna** [reply]
@rdenarie  - As you suggested added the new changes in docker-compose file and added logback file also in same folder of docker compose file., After restart and tried login attempt uploaded the log files. please find the logs file and the configuration files.

$ pwd
**/opt/exo**
[platform.log](https://github.com/exoplatform/exo-community/files/13478263/platform.log)
[docker-compose.txt](https://github.com/exoplatform/exo-community/files/13478265/docker-compose.txt)

$ pwd
**/opt/exo/conf**
[logback.xml.txt](https://github.com/exoplatform/exo-community/files/13478319/logback.xml.txt)

$ pwd
**/opt/exo/gatein/conf**
[exo.txt](https://github.com/exoplatform/exo-community/files/13478379/exo.txt)

Regards,
Raja

**Reply by rdenarie** [reply]
Hello, 
I'm sorry, the docker-compose file you provide do not contains the line for mounting the logback file.
Can you add it ?

**Reply by Rajamuthukrishna** [reply]
@rdenarie - Sorry - I have added the line in docker-compose file and deployed but while attached this portal wrongly uploaded previous file. please find the docker-compose file for added the line.

$ pwd
**/opt/exo**
[docker-compose.txt](https://github.com/exoplatform/exo-community/files/13484912/docker-compose.txt)

**Reply by rdenarie** [reply]
Ok, thanks.
When you go inside the exo container (with docker exec command), can you check that the logback file (/opt/exo/conf/logback.xml) correctly contains the modification you add ?

**Reply by Rajamuthukrishna** [reply]
@rdenarie - Please find the attached screenshots.

<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/501f20ba-ffc9-43c7-b8dc-1323ce01b0fb">
<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/a8bcec0b-4101-41f3-b0b8-1dbb26a38b61">

**Reply by rdenarie** [reply]
Hi @Rajamuthukrishna 
I afraid I made a mistake.
The correct logger is 
`    <logger name="org.gatein.security.oauth.openid" level="TRACE" />`
and not 
`  <logger name="io.meeds.oauth.openid" level="TRACE" />`

Can you change in logback.xml file and restart ?

Then, can you try to log with open id, and then send me the log file ?

Thanks

**Reply by Rajamuthukrishna** [reply]
Hi @rdenarie ,
As added those line in logback.xml file and restarted, getting some below issue while starting exo-community.,
<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/0a2690a9-3fb6-45a5-8a32-25a9f09a12ae">

$ pwd
/opt/exo
[docker-compose.txt](https://github.com/exoplatform/exo-community/files/13499614/docker-compose.txt)

$ pwd
/opt/exo/conf/
[logback.xml.txt](https://github.com/exoplatform/exo-community/files/13499633/logback.xml.txt)

**Reply by rdenarie** [reply]
On your host machine, where did you put the logback.xml file ? 
Just right to the docker-compose file ? or in a folder named 'conf', which is right to the docker compose ?

**Reply by Rajamuthukrishna** [reply]
@rdenarie  - I have placed logback.xml in **/opt/exo**  path and added the logger, and  logback.xml file already there in  **/opt/exo/conf** and updated the logger here aswell.

Added the logback.xml file path in volume of docker-compose in **/opt/exo** path and added the highlighted line.
volumes:
      - exo_data:/srv/exo
      - exo_logs:/var/log/exo
      **- ./logback.xml:/opt/exo/conf/logback.xml**

**Reply by rdenarie** [reply]
/opt/exo/conf/logback.xml is the path INSIDE the exo docker container.

Can you provide the exact path of your docker-compose file in your local machine (not the path inside the docker container, but the real, physical path) ?
Can you also provide the path of the logback file in your local machine  (not the path inside the docker container, but the real, physical path) ?

**Reply by Rajamuthukrishna** [reply]
@rdenarie  - I have updated the my local machine path of logback.xml in docker-compose file. please verify..

**logback.xml** : C:\app\EXO_HOME\conf\logback.xml
[logback.xml.txt](https://github.com/exoplatform/exo-community/files/13501391/logback.xml.txt)

**docker-compose.yml** : C:\app\EXO_HOME\docker-compose.yml
[docker-compose.txt](https://github.com/exoplatform/exo-community/files/13501383/docker-compose.txt)

please find the downloaded log file from the the path of the container in below,
$ pwd
**/opt/exo/logs**
[platform.log](https://github.com/exoplatform/exo-community/files/13501521/platform.log)

**Reply by rdenarie** [reply]
As your logback file is in conf folder, your volume line in docker-compose must be 
      - ./**conf**/logback.xml:C:/app/EXO_HOME/conf/logback.xml:rw

**Reply by Rajamuthukrishna** [reply]
@rdenarie  - After adding this below line and execute the docker-compose file in local - getting too many colons exception...
**docker-compose.yml**
volumes:
      - exo_data:/srv/exo
      - exo_logs:/var/log/exo
      **- ./conf/logback.xml:C:/app/EXO_HOME/conf/logback.xml:rw**

<img width="524" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/9c3bc6be-7e52-4dbe-9803-177944bbeede">

**Reply by rdenarie** [reply]
My mistake : 
- ./conf/logback.xml:/opt/exo/conf/logback.xml:rw

**Reply by Rajamuthukrishna** [reply]
@rdenarie  - Please find the logs...
**docker-compose.yml:**
volumes:
      - exo_data:/srv/exo
      - exo_logs:/var/log/exo
      **- ./conf/logback.xml:/opt/exo/conf/logback.xml:rw**
  
Logs:
[platform.log](https://github.com/exoplatform/exo-community/files/13503028/platform.log)

<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/a6b86c58-0f0f-455a-8230-a8a0d35451ae">

**Comment by rdenarie** [comment]
Ok, your properties for openid seems not taken in account.
As you do for logback file, can you provide the exact of the file exo.properties on your local machine ?

You should have line like this in your docker compose file : 
      - ./conf/exo.properties:/opt/exo/gatein/conf/exo.properties:rw

**Reply by Rajamuthukrishna** [reply]
@rdenarie  - After adding this below line, it will create folder in exo.properties
**docker-compose.yml:**
volumes:
      - exo_data:/srv/exo
      - exo_logs:/var/log/exo
      **- ./conf/exo.properties:/opt/exo/gatein/conf/exo.properties:rw**

<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/0d6e6b04-8655-4e1b-bdb6-06d2e4449d87">

**exo.properties:**
[exo.txt](https://github.com/exoplatform/exo-community/files/13503437/exo.txt)

**Reply by rdenarie** [reply]
Please provide your docker compose file

**Reply by Rajamuthukrishna** [reply]
@rdenarie  - Please find the docker-compose file.
[docker-compose.txt](https://github.com/exoplatform/exo-community/files/13503681/docker-compose.txt)

**Reply by rdenarie** [reply]
Xhat is the exact path or the file exo.properties on your local machine

**Reply by Rajamuthukrishna** [reply]
@rdenarie  - Please find the path of the exo.properties in local machine.
**C:\app\EXO_HOME\gatein\conf\exo.properties**
<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/d5370e47-1357-437f-bb96-bc4a66d3351a">

**Reply by rdenarie** [reply]
So your path in docker-compose should be adapted.
you put ./conf/exo.properties, but it should be 
./gatein/conf/exo.properties :
`      - ./gatein/conf/exo.properties:/opt/exo/gatein/conf/exo.properties:rw`

**Reply by Rajamuthukrishna** [reply]
@rdenarie -Please find the logs..
[platform.log](https://github.com/exoplatform/exo-community/files/13509551/platform.log)

**docker-compose.yml:**
volumes:
      - exo_data:/srv/exo
      - exo_logs:/var/log/exo
      **- ./gatein/conf/exo.properties:/opt/exo/gatein/conf/exo.properties:rw**

**Reply by rdenarie** [reply]
Hello
I can see this in log 
`2023-11-30 06:56:01,704 | ERROR | Connection refused [o.g.s.oauth.openid.OpenIdProcessorImpl<main>] `

I'm afraid that your exo docker container cannot contact the IDP, he didnt manage to read the well-known url.

Please ensure that communication between exo docker container and IDP is possible (connect inside the container, and check with ping, wget ....)

It is problem related to your docker installation, on which I will not be able to help you. 
So I let you check and get back when you are sure that link between eXo and IDP is ok.

Thanks

**Reply by Rajamuthukrishna** [reply]
Tried multiple time for uninstall and install the docker and tried to connect well-known url in docker container getting below details..,Any idea?
$ pwd
/opt/exo
$ wget http://localhost:8090/auth/realms/master/.well-known/openid-configuration
<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/8dd18c73-afba-4056-95cd-083e6284bda1">

Can you pls confirm the exo.properties configuration..
[exo.txt](https://github.com/exoplatform/exo-community/files/13511086/exo.txt)

**Comment by rdenarie** [comment]
The exo properties seems ok.

in fact, in exo container, when you ping  or wget localhost, you address the exo container. 

As your idp seems running on your local, physical machine, I think you need to provide the IP of your machine as IDP host name

**Reply by Rajamuthukrishna** [reply]
Tried to execute the container **wget http://127.0.0.1:8090/auth/realms/master/.well-known/openid-configuration** but getting connection refused exception.
<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/062a3a86-a92b-46d8-9d0f-979aa3f23b96">

**Reply by rdenarie** [reply]
127.0.0.1 is exactly the same as localhost.

You should use your local machine IP

**Reply by Rajamuthukrishna** [reply]
I have tried my machine IP - getting connection refused.
<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/4fe216d7-5422-4458-b429-c2ae96cf8f2d">

even, if we use machine IP also- we are getting localhost in the response. please find the logs..
<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/937c7ce9-3cae-4eeb-8265-004b471ff6ad">

**Reply by rdenarie** [reply]
In fact, it seems related to your docker installation and machine related. 

As there is no link with exo platform,  I will not be able to help you on this point

**Reply by Rajamuthukrishna** [reply]
@rdenarie Yes - You are correct. The issue is from my local keyclock only.

I just use different keyclock, able to connect and redirect the keyclock page..but getting some url issue
<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/dd7bf061-587a-4b9c-97dc-200a865590a8">
<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/b642b6d3-4e14-42b7-9996-65e0ee853e3d">

<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/191e57e1-4843-4858-b403-443231268140">

[platform.log](https://github.com/exoplatform/exo-community/files/13516927/platform.log)

**Reply by rdenarie** [reply]
Invalid parameter redirect_uri generally means that you didnt register your service provider in the IDP.

When using Google as IDP, you haev to configure which redirect uri are authorized : 
![image](https://github.com/exoplatform/exo-community/assets/807839/41fab2b8-b645-44e5-b267-ea7ff188d19f)

In your case you shouln't have done this step

**Reply by Rajamuthukrishna** [reply]
I haven't provided the redirectUrl in keyclock IDP but still getting some issue. I think it will refer from exo.prop only.
<img width="613" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/d8e04837-9862-4057-87f1-b243148a6d3c">

If I have commented the redirectUrl in eXo.properties getting below exception., Can you pls confirm any proper url will provide in rediectURL in exo.props.
<img width="721" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/19b4857c-3fb3-4d46-ba60-b5547ebb67ab">
<img width="960" alt="image" src="https://github.com/exoplatform/exo-community/assets/112479162/64861649-6bcf-47cc-9fbe-a3b7393ad6db">

**Comment by rdenarie** [comment]
redirect_uri is mandatory in exo.properties.
It is also mandatory in Keycloack configuration

**Reply by Rajamuthukrishna** [reply]
Ok, Can you pls confirm - which redirecturi provide to eXo.prop and keyclock? I have added below uri but not working..
exo.oauth.openid.redirectURL=http://localhost/portal/openidAuth

**Reply by rdenarie** [reply]
Perhaps you should follow these type of questions : https://stackoverflow.com/questions/45352880/keycloak-invalid-parameter-redirect-uri

This one is ok 
exo.oauth.openid.redirectURL=http://localhost/portal/openidAuth

**Reply by Rajamuthukrishna** [reply]
Its Resolved.

@rdenarie -Thank you so much for your kind response. without your support not able to complete this task.

**Comment by rdenarie** [comment]
Thanks for your notice. I close the discussion
